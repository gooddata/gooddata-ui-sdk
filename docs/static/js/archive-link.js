document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});