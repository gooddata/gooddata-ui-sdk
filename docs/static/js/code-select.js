(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));