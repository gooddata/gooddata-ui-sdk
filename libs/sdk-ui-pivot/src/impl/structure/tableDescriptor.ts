// (C) 2021 GoodData Corporation

import {
    AnyCol,
    DataColLeaf,
    isDataColGroup,
    isDataColLeaf,
    isSliceCol,
    isDataColRootGroup,
    SliceCol,
    TableColDefs,
    TableCols,
    agColId,
    DataColGroup,
    isEmptyDataColGroup,
} from "./tableDescriptorTypes";
import { ColDef, ColGroupDef, Column } from "@ag-grid-community/all-modules";
import invariant from "ts-invariant";
import { IAttributeColumnWidthItem, IMeasureColumnWidthItem } from "../../columnWidths";
import { IAttributeDescriptor, IMeasureDescriptor } from "@gooddata/sdk-backend-spi";
import { searchForLocatorMatch } from "./colLocatorMatching";
import { DataViewFacade } from "@gooddata/sdk-ui";
import { createHeadersAndColDefs } from "./tableDescriptorFactory";
import { ISortItem } from "@gooddata/sdk-model";
import { createSortIndicators, SortIndicator } from "./tableDescriptorSorting";
import { createSortItemForCol } from "./colSortItemFactory";
import keyBy from "lodash/keyBy";

/**
 * Table Descriptor is the entry point to all table structure data and metadata. It contains exhaustive information
 * about all table columns (column descriptors) and their respective ag-grid column definitions.
 *
 * On top of this data and metadata, the table descriptor provides convenience and added value methods.
 *
 * Column Descriptors vs ag-grid ColDefs | ColGroupDefs
 * ----------------------------------------------------
 *
 * Column Descriptors (shortened to `Col` sakes of brevity) are our implementation-specific descriptors for
 * the table columns and their grouping into column groups. They contain all the essential GD-specific metadata about
 * the content of the respective table column (attribute descriptors, headers, measure descriptors) and additional
 * structural information.
 *
 * Column Descriptors are divided into two groups: slicing column descriptors and data column descriptors. The
 * slicing columns essentially describe the row attribute columns, while the data columns describe columns that may
 * contains the computed metric values. The data column is a composite. In tables the scope metrics for one or more
 * attributes, the table columns will be further grouped by the different attribute elements to which the metrics
 * were scoped to.
 *
 * The ag-grid ColDefs and ColGroupDefs are (naturally) used to construct the ag-grid table :) They are built to
 * reflect the column descriptors but only contain the information needed by ag-grid itself.
 *
 * The important thing to remember is that ColDefs and ColGroupDefs have same colId/groupId as their column descriptor
 * counterparts.
 *
 * @alpha
 */
export class TableDescriptor {
    /**
     * This field contains slice column descriptors zipped with their respective ColDef that should
     * be used for ag-grid.
     */
    public readonly zippedSliceCols: Array<[SliceCol, ColDef]> = [];

    /**
     * This field contains descriptors of leaf columns zipped with their respective ColDef that should
     * be used for ag-grid.
     */
    public readonly zippedLeaves: Array<[DataColLeaf | DataColGroup, ColDef]> = [];
    private readonly measureColumns: number;

    private constructor(
        private readonly dv: DataViewFacade,
        public readonly headers: TableCols,
        public readonly colDefs: TableColDefs,
    ) {
        this._initializeZippedLeaves();
        this._initializeZippedSliceCols();
        this.measureColumns = headers.leafDataCols.filter(isDataColLeaf).length;
    }

    /**
     * Creates a new table descriptor from the provided data view facade.
     *
     * @param dv - data view facade
     */
    public static for(dv: DataViewFacade): TableDescriptor {
        const { headers, colDefs } = createHeadersAndColDefs(dv);

        invariant(headers.leafDataCols.length === colDefs.leafDataColDefs.length);

        return new TableDescriptor(dv, headers, colDefs);
    }

    private _initializeZippedLeaves() {
        this.headers.leafDataCols.forEach((col, idx) => {
            this.zippedLeaves.push([col, this.colDefs.leafDataColDefs[idx]]);
        });
    }

    private _initializeZippedSliceCols() {
        this.headers.sliceCols.forEach((col, idx) => {
            this.zippedSliceCols.push([col, this.colDefs.sliceColDefs[idx]]);
        });
    }

    /**
     * Gets column descriptor for column with the provided ID. This method will raise error if there is no such
     * col: this is to draw out possible errors in the rest of table code that uses the cols and ColDefs.
     *
     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
     */
    public getCol(c: string | Column | ColDef | ColGroupDef): AnyCol {
        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
        const id = agColId(c);
        const result = this.headers.idToDescriptor[id];

        invariant(result, `no column with id ${id}`);

        return result;
    }

    /**
     * Gets ColDef or ColGroupDef for column with the provided ID. This method will raise error if there is no such
     * col: this is to draw out possible errors in the rest of table code that uses the cols and ColDefs.
     *
     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
     */
    public getColDef(c: string | Column | ColDef | ColGroupDef): ColDef | ColGroupDef {
        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
        const id = agColId(c);
        const result = this.colDefs.idToColDef[id];

        invariant(result, `no column with id ${id}`);

        return result;
    }

    /**
     * Gets descriptors of all attributes that are used to slice the table into rows. Note that it is perfectly
     * OK that table has no slicing attributes.
     *
     * @returns empty if there are no slicing attributes
     */
    public getSlicingAttributes(): IAttributeDescriptor[] {
        return this.headers.sliceCols.map((col) => col.attributeDescriptor);
    }

    /**
     *
     */
    public sliceColCount(): number {
        return this.headers.sliceCols.length;
    }

    /**
     * Gets all descriptors of all measures that were used to compute data values for this table. Note that it is
     * perfectly OK that table has no measures. If the table has slicing attributes, then it will be listing out
     * all available attribute values.
     *
     * Also note that table MAY have more data leaf columns than there are number of measures. If the measures
     * are further scoped for values of some attribute's elements, then there will be one data col for each combination
     * of measure X attribute element.
     *
     * @returns empty if there are no measures
     */
    public getMeasures(): IMeasureDescriptor[] {
        return this.dv.meta().measureDescriptors();
    }

    /**
     * Returns count of leaf data cols. This represents the actual width of the data sheet holding the computed metric
     * values.
     */
    public dataColLeavesCount(): number {
        return this.measureColumns;
    }

    /**
     * Tests whether the column with the provided id is the first (e.g. left-most) column in the table. Table always starts with
     * slicing columns followed by data columns.
     *
     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
     */
    public isFirstCol(c: string | Column | ColDef | ColGroupDef): boolean {
        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
        const id = agColId(c);

        if (this.headers.sliceCols.length > 0) {
            return this.headers.sliceCols[0].id === id;
        }

        return this.isFirstDataCol(id);
    }

    /**
     * Tests whether the column with the provided id is the first (e.g. left-most) column of the data sheet part of the table.
     *
     * Note that for table that uses column groups, there will be multiple first columns: the grouping root and recursively
     * its first children down to the first leaf col..
     *
     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
     */
    public isFirstDataCol(c: string | Column | ColDef | ColGroupDef): boolean {
        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
        const col = this.getCol(c);

        switch (col.type) {
            case "leafColumnDescriptor":
            case "columnGroupHeaderDescriptor": {
                return col.fullIndexPathToHere.every((idx) => idx === 0);
            }
            case "columnGroupRootDescriptor": {
                return true;
            }
            default: {
                return false;
            }
        }
    }

    /**
     * Tests whether the table has grouped data columns.
     */
    public hasGroupedDataCols(): boolean {
        const firstRoot = this.headers.rootDataCols[0];

        return firstRoot && (isDataColGroup(firstRoot) || isDataColRootGroup(firstRoot));
    }

    /**
     * Tests whether the table has any leaf data cols - in other words whether there any computed data values to show
     * in the table. It is OK for table not to have any data leaf cols - it may be just a table with slicing cols listing
     * elements of an attribute.
     */
    public hasDataLeafCols(): boolean {
        return this.headers.leafDataCols.length > 0;
    }

    /**
     * Given a column that may appear as a leaf of table headers this method returns its absolute index in the table.
     *
     * This takes into account that the table columns go from left-to-right, starting with slicing columns first then
     * followed by leaf data columns.
     *
     * @param col - column to get absolute index of
     */
    public getAbsoluteLeafColIndex(col: SliceCol | DataColLeaf | DataColGroup): number {
        if (isSliceCol(col)) {
            return col.index;
        } else if (isDataColGroup(col)) {
            // if this bombs, caller is not operating with the leaf columns correctly and sent over
            // a col that is not a leaf
            invariant(isEmptyDataColGroup(col));

            return this.sliceColCount() + this.headers.leafDataCols.findIndex((leaf) => leaf.id === col.id);
        }

        return this.sliceColCount() + col.index;
    }

    /**
     * Attempts to match the provided attribute column width item to a SliceCol descriptor.
     *
     * @param columnWidthItem - item to match
     */
    public matchAttributeWidthItem(columnWidthItem: IAttributeColumnWidthItem): SliceCol | undefined {
        const matcher = attributeDescriptorLocalIdMatch(
            columnWidthItem.attributeColumnWidthItem.attributeIdentifier,
        );

        return this.headers.sliceCols.find((col) => matcher(col.attributeDescriptor));
    }

    /**
     * Attempts to match the provided measure width item to a leaf data col. The locators in the item
     * will be used to traverse the column structure.
     *
     * @param measureWidthItem - item to match
     */
    public matchMeasureWidthItem(
        measureWidthItem: IMeasureColumnWidthItem,
    ): DataColLeaf | DataColGroup | undefined {
        return searchForLocatorMatch(
            this.headers.rootDataCols,
            measureWidthItem.measureColumnWidthItem.locators,
        );
    }

    /**
     * Tests whether the table can be enriched by totals. Tables that do not have any measures or do not have any
     * slicing attributes cannot have totals. Because by definition they either have exactly 1 row with all measure grant total
     * sum or have no rows whatsover.
     */
    public canTableHaveTotals(): boolean {
        return this.sliceColCount() > 0 && this.dataColLeavesCount() > 0;
    }

    /**
     * Returns slice col against which grant totals can be defined. This is essentially the left-most slicing column.
     * This method will raise invariant error if there are no slicing cols. Rationale being that inclusion of grand
     * totals requires more checks anyway and calling this blindly does not make much sense.
     *
     * @remarks see {@link TableDescriptor.canTableHaveTotals}
     */
    public getGrandTotalCol(): SliceCol {
        const result = this.headers.sliceCols[0];

        invariant(result);

        return result;
    }

    /**
     * Given a col, this function determines all slice cols that are effective up to and including
     * the col. In other words, it returns all slice cols 'to the left' of the provided col.
     *
     * Note: the col can be of any type. If a data col is provided, then by definition all slice cols are
     * effective. if col is a slice col, then slice cols from start up to and including the provided col
     * are returned.
     */
    public getSliceColsUpToIncludingCol(col: AnyCol): SliceCol[] {
        const allSliceCols = this.headers.sliceCols;

        if (!isSliceCol(col)) {
            return allSliceCols;
        }

        const colAt = allSliceCols.findIndex((slice) => slice.id === col.id);

        // if this happens, then caller has mismatch of descriptor and cols
        invariant(colAt > -1);

        return allSliceCols.slice(0, colAt + 1);
    }

    /**
     * Given ag-grid columns or coldefs and a list of initial sort items, this function will create
     * sort items that reflect current table sorting configuration. The sorting configuration is obtained from the
     * Column/ColDef `getSort`/`sort` property.
     *
     * Note: the initial sort items are passed to here to ensure that attribute area sort setting is reused
     * correctly when sort direction for the respective column changes.
     *
     * @param columns - ag-grid columns / coldefs to obtain sorting information from
     * @param originalSorts - original sorts
     */
    public createSortItems(
        columns: Array<Column> | Array<ColDef>,
        originalSorts: ISortItem[] = [],
    ): ISortItem[] {
        if (columns.length === 0) {
            return [];
        }

        const sortIndicators = createSortIndicators(columns);

        return sortIndicators.map((sortModelItem: SortIndicator) => {
            const { colId, sort } = sortModelItem;
            const col = this.getCol(colId);

            return createSortItemForCol(col, sort, originalSorts);
        });
    }

    /**
     * Updates effective totals for the slice cols using the new total descriptors included for their respective
     * attributes in the new data view facade.
     *
     * @param dv - data view with same attribute structure but with added totals
     */
    public updateEffectiveTotals(dv: DataViewFacade): void {
        const idToDescriptor: Record<string, IAttributeDescriptor> = keyBy(
            dv.meta().attributeDescriptors(),
            (desc) => desc.attributeHeader.localIdentifier,
        );

        this.headers.sliceCols.forEach((sliceCol) => updateEffectiveTotals(sliceCol, idToDescriptor));
    }
}

function attributeDescriptorLocalIdMatch(localId: string): (b: IAttributeDescriptor) => boolean {
    return (b: IAttributeDescriptor): boolean => {
        return localId === b.attributeHeader.localIdentifier;
    };
}

function updateEffectiveTotals(col: SliceCol, newDescriptors: Record<string, IAttributeDescriptor>) {
    const attributeLocalId = col.attributeDescriptor.attributeHeader.localIdentifier;
    const newDescriptor = newDescriptors[attributeLocalId];

    // if this bombs then reinit logic of the entire pivot table is flawed because upon change of table structure is
    // re-initialized - which includes the table descriptor so the code should be getting to this place at all.
    invariant(
        newDescriptor,
        `attempting to refresh attribute descriptors for different table. attribute with local id ${attributeLocalId} not found`,
    );

    col.effectiveTotals = newDescriptor.attributeHeader.totalItems ?? [];
}
