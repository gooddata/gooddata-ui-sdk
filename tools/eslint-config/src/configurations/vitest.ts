// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
