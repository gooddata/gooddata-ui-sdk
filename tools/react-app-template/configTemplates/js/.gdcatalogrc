{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}