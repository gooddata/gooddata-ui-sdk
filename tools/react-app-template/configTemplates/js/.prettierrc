{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
