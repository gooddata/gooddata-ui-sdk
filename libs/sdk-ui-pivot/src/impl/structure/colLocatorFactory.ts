// (C) 2021-2022 GoodData Corporation

import { isScopeCol, LeafDataCol, TransposedMeasureDataCol } from "./tableDescriptorTypes.js";
import {
    ColumnLocator,
    ITotalColumnLocator,
    IAttributeColumnLocator,
    IMeasureColumnLocator,
} from "../../columnWidths.js";
import {
    IMeasureDescriptor,
    IAttributeDescriptor,
    IResultAttributeHeader,
    isResultTotalHeader,
    IResultTotalHeader,
} from "@gooddata/sdk-model";
import { invariant } from "ts-invariant";
import zip from "lodash/zip.js";

function createAttributeLocator(
    descriptor: IAttributeDescriptor | undefined,
    header: IResultAttributeHeader | undefined,
): IAttributeColumnLocator {
    // if this bombs it means something is wrong with the col descriptors or in very bad case in the DVF data access logic.
    // by contract, all data series have same number of descriptors & headers. therefore the zipping logic should never
    // run into situation where the arrays are of different size.
    invariant(descriptor && header);

    return {
        attributeLocatorItem: {
            attributeIdentifier: descriptor.attributeHeader.localIdentifier,
            element: header.attributeHeaderItem.uri,
        },
    };
}

function createTotalLocator(
    descriptor: IAttributeDescriptor | undefined,
    header: IResultTotalHeader | undefined,
): ITotalColumnLocator {
    // if this bombs it means something is wrong with the col descriptors or in very bad case in the DVF data access logic.
    // by contract, all data series have same number of descriptors & headers. therefore the zipping logic should never
    // run into situation where the arrays are of different size.
    invariant(descriptor && header);

    return {
        totalLocatorItem: {
            attributeIdentifier: descriptor.attributeHeader.localIdentifier,
            totalFunction: header.totalHeaderItem.name,
        },
    };
}

function createMeasureLocator(descriptor: IMeasureDescriptor): IMeasureColumnLocator {
    return {
        measureLocatorItem: {
            measureIdentifier: descriptor.measureHeaderItem.localIdentifier,
        },
    };
}

/**
 * Given a leaf data col, this function will create column locator that can be used in
 * width items.
 *
 * @param col - col definition to get locators for
 */
export function createColumnLocator(col: LeafDataCol): ColumnLocator[] {
    if (isScopeCol(col)) {
        const result: ColumnLocator[] = [];
        const { descriptorsToHere, headersToHere, measureDescriptors } = col;
        const descriptorsAndHeaders = zip(descriptorsToHere, headersToHere);
        descriptorsAndHeaders.push([col.attributeDescriptor, col.header]);

        descriptorsAndHeaders.forEach(([descriptor, header]) =>
            isResultTotalHeader(header)
                ? result.push(createTotalLocator(descriptor, header))
                : result.push(createAttributeLocator(descriptor, header)),
        );

        if (measureDescriptors) {
            measureDescriptors.forEach((measureDescriptor) =>
                result.push(createMeasureLocator(measureDescriptor)),
            );
        }

        return result;
    } else {
        const result: ColumnLocator[] = [];

        if (col.seriesDescriptor.attributeDescriptors && col.seriesDescriptor.attributeHeaders) {
            const descriptorAndHeaders = zip(
                col.seriesDescriptor.attributeDescriptors,
                col.seriesDescriptor.attributeHeaders,
            );

            descriptorAndHeaders.forEach(([descriptor, header]) => {
                if (isResultTotalHeader(header)) {
                    result.push(createTotalLocator(descriptor, header));
                } else {
                    result.push(createAttributeLocator(descriptor, header));
                }
            });
        }

        result.push(createMeasureLocator(col.seriesDescriptor.measureDescriptor));

        return result;
    }
}

/**
 * Given a transposed slice measure or mixed values data col, this function will create column locator that can be used in
 * width items.
 *
 * @param col - col definition to get locators for
 */
export function createTransposedColumnLocator(
    col: TransposedMeasureDataCol | TransposedMeasureDataCol,
): IMeasureColumnLocator[] {
    const result: IMeasureColumnLocator[] = [];

    if (col.seriesDescriptor) {
        const filteredSeriesDescriptor = col.seriesDescriptor.filter(
            (serieDescriptor, index, self) =>
                self.findIndex((s) => s.measureDescriptor === serieDescriptor.measureDescriptor) === index,
        );

        filteredSeriesDescriptor.forEach((serieDescriptor) => {
            result.push(createMeasureLocator(serieDescriptor.measureDescriptor));
        });
    }

    return result;
}
