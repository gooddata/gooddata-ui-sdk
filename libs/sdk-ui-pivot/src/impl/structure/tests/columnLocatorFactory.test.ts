// (C) 2007-2021 GoodData Corporation

import {
    MultipleMeasuresAndNoColumnsWithMetricsInRows,
    SingleMeasureWithRowAttribute,
    SingleMeasureWithTwoRowAndTwoColumnAttributes,
    TwoMeasuresWithSingleRowAttrWithMetricsInRows,
} from "./table.fixture.js";
import { TableDescriptor } from "../tableDescriptor.js";
import { createColumnLocator, createTransposedColumnLocator } from "../colLocatorFactory.js";
import { ScopeCol, SliceMeasureCol, MixedValuesCol } from "../tableDescriptorTypes.js";
import { describe, it, expect } from "vitest";

describe("createColumnLocator", () => {
    it("creates valid leaf column locator in table without column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithRowAttribute, "empty value");

        expect(createColumnLocator(t.headers.leafDataCols[0])).toMatchSnapshot();
    });

    it("creates valid leaf column locator in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes, "empty value");

        expect(createColumnLocator(t.headers.leafDataCols[0])).toMatchSnapshot();
    });

    it("creates valid group column locator for top-level group in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes, "empty value");

        const topLevelGroup = t.headers.rootDataCols[0].children[0] as ScopeCol;

        expect(createColumnLocator(topLevelGroup)).toMatchSnapshot();
    });

    it("creates valid group column locator for second-level group in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes, "empty value");

        const secondLevelGroup = t.headers.rootDataCols[0].children[0].children[0] as ScopeCol;

        expect(createColumnLocator(secondLevelGroup)).toMatchSnapshot();
    });

    it("creates valid slice measure column locator in transposed table", () => {
        const t = TableDescriptor.for(TwoMeasuresWithSingleRowAttrWithMetricsInRows, "empty value");

        const sliceMeasureCol = t.headers.sliceMeasureCols[0] as SliceMeasureCol;

        expect(createTransposedColumnLocator(sliceMeasureCol)).toMatchSnapshot();
    });

    it("creates valid mixed values column locator in transposed table", () => {
        const t = TableDescriptor.for(MultipleMeasuresAndNoColumnsWithMetricsInRows, "empty value");

        const mixedValuesCol = t.headers.mixedValuesCols[0] as MixedValuesCol;

        expect(createTransposedColumnLocator(mixedValuesCol)).toMatchSnapshot();
    });
});
