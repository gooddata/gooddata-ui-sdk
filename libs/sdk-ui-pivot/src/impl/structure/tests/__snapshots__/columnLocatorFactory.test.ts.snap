// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createColumnLocator > creates valid group column locator for second-level group in table with column attributes 1`] = `
[
  {
    "attributeLocatorItem": {
      "attributeIdentifier": "a_attr.f_stage.stagename",
      "element": "Interest",
    },
  },
  {
    "attributeLocatorItem": {
      "attributeIdentifier": "a_f_owner.region_id",
      "element": "East Coast",
    },
  },
]
`;

exports[`createColumnLocator > creates valid group column locator for top-level group in table with column attributes 1`] = `
[
  {
    "attributeLocatorItem": {
      "attributeIdentifier": "a_attr.f_stage.stagename",
      "element": "Interest",
    },
  },
]
`;

exports[`createColumnLocator > creates valid leaf column locator in table with column attributes 1`] = `
[
  {
    "attributeLocatorItem": {
      "attributeIdentifier": "a_attr.f_stage.stagename",
      "element": "Interest",
    },
  },
  {
    "attributeLocatorItem": {
      "attributeIdentifier": "a_f_owner.region_id",
      "element": "East Coast",
    },
  },
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
    },
  },
]
`;

exports[`createColumnLocator > creates valid leaf column locator in table without column attributes 1`] = `
[
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
    },
  },
]
`;

exports[`createColumnLocator > creates valid mixed values column locator in transposed table 1`] = `
[
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
    },
  },
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
    },
  },
]
`;

exports[`createColumnLocator > creates valid slice measure column locator in transposed table 1`] = `
[
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
    },
  },
  {
    "measureLocatorItem": {
      "measureIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
    },
  },
]
`;
